### PR TITLE
Rework LUT to batch processing and remove access to TL buffer

### PR DESCRIPTION
**Pull request by klecki**
## Description
- [ ] **Bug fix** (*non-breaking change which fixes an issue*)
- [ ] **New feature** (*non-breaking change which adds functionality*)
- [ ] **Breaking change** (*fix or feature that would cause existing functionality to not work as expected*)
- [x] **Refactoring** (*Redesign of existing code that doesn't affect functionality*)
- [ ] **Other** (*e.g. Documentation, Tests, Configuration*)

### What happened in this PR
Batch processing introduced for CPU implementation.
One copy of LUT table kept for GPU implementation.
Remove access to underlying contiguous TL buffer in the op
by implementing per-sample CUDA kernel.
 

#### Additional information
- Affected modules and functionalities:
LUT

- Key points relevant for the review:
N/A

### Tests
- [x] Existing tests apply
- [ ] New tests added
  - [ ] Python tests
  - [ ] GTests
  - [ ] Benchmark
  - [ ] Other
- [ ] N/A

### Documentation
- [x] Existing documentation applies
- [ ] Documentation updated
  - [ ] Docstring
  - [ ] Doxygen
  - [ ] RST
  - [ ] Jupyter
  - [ ] Other
- [ ] N/A

#### Requirements
- [ ] Implements new requirements
- [ ] Affects existing requirements
- [x] N/A

**REQ IDs**: N/A

**JIRA TASK**: N/A

### CODE REVIEW

**Comment by klecki** [comment]
!build

**Comment by klecki** [comment]
!build

File: dali/operators/generic/lookup_table.cc [hunk 1]
```diff
@@ -30,21 +33,32 @@ void LookupTable<CPUBackend>::RunImpl(SampleWorkspace &ws) {
...
+                if (check_range) {
+                  out_data[i] = (std::is_unsigned<InputType>::value || key >= 0) && key <= max_key ?
+                                    lookup_table[key] :
+                                    default_value;
+                } else {
+                  out_data[i] = lookup_table[key];
+                }
+              }
+            },
+            data_size);
```
Comment (mzient) [inline dali/operators/generic/lookup_table.cc:56]: Move this to a function, perhaps?

File: dali/operators/generic/lookup_table.cu [hunk 2]
```diff
@@ -24,57 +25,67 @@ namespace detail {
...
 
-  const auto key = input[tid];
-  if (check_range) {
-    output[tid] = (std::is_unsigned<InputType>::value || key >= 0) && key <= max_key ?
-      lookup_table[key] : default_value;
-  } else {
-    output[tid] = lookup_table[key];
+  auto *output = reinterpret_cast<OutputType *>(sample.output);
+  const auto *input = reinterpret_cast<const InputType *>(sample.input);
+  for (int x = threadIdx.x + block.start.x; x < block.end.x; x += blockDim.x) {
```
Comment (mzient) [inline dali/operators/generic/lookup_table.cu:40]: I have some reservations for the use of BlockSetup here. It was initially meant to work with multidimensional data and does not support tensors in excess of 2G in each dimension - not a problem for large spatial problems, but for a flattened tensor we could hit this limit pretty easily and face dire consequences.

File: dali/operators/generic/lookup_table.cu [hunk 3]
```diff
@@ -24,57 +25,67 @@ namespace detail {
...
+  auto num_samples = shape.num_samples();
+  samples_.resize(num_samples);
+  for (int sample_id = 0; sample_id < num_samples; sample_id++) {
+    samples_[sample_id].output = output.raw_mutable_tensor(sample_id);
+    samples_[sample_id].input = input.raw_tensor(sample_id);
+  }
+  samples_dev_.from_host(samples_, stream);
+
+  std::array<std::pair<int, int>, 1> collapse_groups = {{{0, shape.sample_dim()}}};
+  auto collapsed_shape = collapse_dims<1>(shape, collapse_groups);
```
Comment (mzient) [inline dali/operators/generic/lookup_table.cu:72]: Nitpick - I guess you could get away with:
```suggestion
  auto collapsed_shape = collapse_dims<1>(shape, { make_pair(0, shape.sample_dim()) });
```

File: dali/operators/generic/lookup_table.cu (hunk 2 above)
Comment (klecki) [reply dali/operators/generic/lookup_table.cu:40]: Addressed here: https://github.com/NVIDIA/DALI/pull/3304

File: dali/operators/generic/lookup_table.cc (hunk 1 above)
Comment (klecki) [reply dali/operators/generic/lookup_table.cc:56]: Done

File: dali/operators/generic/lookup_table.cu (hunk 3 above)
Comment (klecki) [reply dali/operators/generic/lookup_table.cu:72]: done

**Comment by klecki** [comment]
!build

File: dali/operators/generic/lookup_table.cu (hunk 2 above)
Comment (mzient) [reply dali/operators/generic/lookup_table.cu:33]:
```suggestion
  for (auto x = threadIdx.x + block.start.x; x < block.end.x; x += blockDim.x) {
```

**Comment by klecki** [comment]
!build

**Review by mzient — APPROVED** [review]

File: dali/operators/generic/lookup_table.h [hunk 4]
```diff
@@ -41,20 +45,27 @@ void value_mem_deleter(void *ptr) {
...
   explicit inline LookupTable(const OpSpec &spec)
-    : Operator<Backend>(spec)
-    , input_type_(DALI_NO_TYPE)
-    , output_type_(spec.GetArgument<DALIDataType>("dtype"))
-    , default_value_f_(spec.GetArgument<float>("default_value")) {
+      : Operator<Backend>(spec),
+        input_type_(DALI_NO_TYPE),
+        output_type_(spec.GetArgument<DALIDataType>("dtype")),
+        default_value_f_(spec.GetArgument<float>("default_value")) {
+    lut_.Reset();
```
Comment (szalpal) [inline dali/operators/generic/lookup_table.h:64]: 👍🏻

**Review by szalpal — APPROVED** [review]

File: dali/operators/generic/lookup_table.cc [hunk 5]
```diff
@@ -14,37 +14,52 @@
...
+
 namespace dali {
 
-template<>
-void LookupTable<CPUBackend>::RunImpl(SampleWorkspace &ws) {
-  const auto &input = ws.Input<CPUBackend>(0);
-  auto &output = ws.Output<CPUBackend>(0);
+namespace {
+
+template <typename Output, typename Input>
```
[the remark this answers is not in the document]
Comment (klecki) [reply dali/operators/generic/lookup_table.cc:22]: Yeah, I kept the Output, Input order, but I had similar feeling about inferring one of the types from arguments.

**Pull request merged**